sklearn/partial_dependence.py

```diff
@@ -0,0 +1,577 @@
+
+    if est.init is not None:
+        warnings.warn(
+            'Using recursion method with a non-constant init predictor will '
+            'lead to incorrect partial dependence values.',
+            UserWarning
+        )
+
+    # grid needs to be DTYPE
+    grid = np.asarray(grid, dtype=DTYPE, order='C')
+
+    n_trees_per_stage = est.estimators_.shape[1]
+    n_estimators = est.estimators_.shape[0]
+    learning_rate = est.learning_rate
+    averaged_predictions = np.zeros((n_trees_per_stage, grid.shape[0]),
+                                    dtype=np.float64, order='C')
+    for stage in range(n_estimators):
+        for k in range(n_trees_per_stage):
+            tree = est.estimators_[stage, k].tree_
+            _partial_dependence_tree(tree, grid, target_variables,
+                                     learning_rate, averaged_predictions[k])
+
+    return averaged_predictions
+
+
+def _partial_dependence_brute(est, grid, target_variables, X):
+
+    averaged_predictions = []
+    for new_values in grid:
+        X_eval = X.copy()
+        for i, variable in enumerate(target_variables):
+            X_eval[:, variable] = new_values[i]
+
+        try:
+            predictions = (est.predict(X_eval) if is_regressor(est)
+                           else est.predict_proba(X_eval))
+        except NotFittedError:
+            raise ValueError('est parameter must be a fitted estimator')
+
+        # Note: predictions is of shape
+        # (n_points,) for non-multioutput regressors
+        # (n_points, n_tasks) for multioutput regressors
+        # (n_points, 1) for the regressors in cross_decomposition (I think)
+        # (n_points, 2)  for binary classifaction
+        # (n_points, n_classes) for multiclass classification
+
+        # average over samples
+        averaged_predictions.append(np.mean(predictions, axis=0))
+
+    # reshape to (n_targets, n_points) where n_targets is:
+    # - 1 for non-multioutput regression and binary classification (shape is
+    #   already correct in those cases)
+    # - n_tasks for multi-output regression
+    # - n_classes for multiclass classification.
+    averaged_predictions = np.array(averaged_predictions).T
+    if is_regressor(est) and averaged_predictions.ndim == 1:
+        # non-multioutput regression, shape is (n_points,)
+        averaged_predictions = averaged_predictions.reshape(1, -1)
+    elif is_classifier(est) and averaged_predictions.shape[0] == 2:
+        # Binary classification, shape is (2, n_points).
+        # we output the effect of **positive** class
+        averaged_predictions = averaged_predictions[1]
+        averaged_predictions = averaged_predictions.reshape(1, -1)
+
+    return averaged_predictions
+
+
+def partial_dependence(est, target_variables, X, percentiles=(0.05, 0.95),
+                       grid_resolution=100, method='auto'):
+    """Partial dependence of ``target_variables``.
+
+    .. _warning_recursion_init:
+
+    .. warning::
```

Comment:
I think this belongs in a Notes or Warnings section below, not here in the description of the function.